openverse_catalog/dags/providers/provider_api_scripts/inaturalist.py

```diff
@@ -1,111 +1,161 @@
+    open(os.path.join(SCRIPT_DIR, "05_export_dbpage_to_json_template.sql")).read()
+)
+
+
+class inaturalistDataIngester(ProviderDataIngester):
+
+    # TO DO: consider using this sql to get a max number of db pages to define a
+    # get_should_continue function
+    # select relpages
+    # from pg_class t join pg_namespace s on t.relnamespace=s.oid
+    # where s.nspname='inaturalist' and t.relname='photos';
+
+    providers = {"image": prov.INATURALIST_DEFAULT_PROVIDER}
+
+    def get_next_query_params(self, old_query_params=None, **kwargs):
+        """Page counter"""
+        if old_query_params is None:
+            return {"page_number": 0}
+        else:
+            next_page = old_query_params["page_number"] + 1
+            return {"page_number": next_page}
+
+    def get_response_json(self, query_params: Dict):
+        """
+        Call the SQL to pull json from Postgres, where the raw data has been loaded.
+        """
+        db_page_number = str(query_params["page_number"])
+        sql_string = PG_TO_JSON_TEMPLATE.replace("db_page_number", db_page_number)
+        pg = PostgresHook(POSTGRES_CONN_ID)
+        return pg.get_records(sql_string)
+
+    def get_batch_data(self, response_json):
+        if response_json:
+            return [dict(item[0]) for item in response_json]
+        return None
+
+    def get_record_data(self, data):
+        license_url = data.get("license_url", None)
+        license_info = get_license_info(license_url=license_url)
+        if license_info == LicenseInfo(None, None, None, None):
+            return None
+
+        foreign_id = data.get("foreign_id")
+        if foreign_id is None:
+            return None
+
+        return {
+            "foreign_identifier": f"{foreign_id}",
+            "foreign_landing_url": data.get("foreign_landing_url"),
+            "title": data.get("title"),
+            "creator": data.get("creator"),
+            "image_url": data.get("image_url"),
+            "width": data.get("width"),
+            "height": data.get("height"),
+            "license_info": license_info,
+            "filetype": data.get("filetype"),
+            "creator_url": data.get("creator_url"),
+            "raw_tags": data.get("tags"),
+        }
+
+    def get_media_type(self, record):
+        # This provider only supports Images via S3, though they have some audio files
+        # on site and in the API.
+        return "image"
+
+    def endpoint(self):
+        assert "Postgres DB" == "API endpoint"
+
+
+def sql_runner(file_name):
+    assert os.path.exists(os.path.join(SCRIPT_DIR, file_name))
+    assert file_name[-4:] == ".sql"
+    query = dedent(open(os.path.join(SCRIPT_DIR, file_name), "r").read())
+    pg = PostgresHook(POSTGRES_CONN_ID)
+    return pg.get_records(query)
+
+
+# defining separate python callables for separate airflow tasks.
+
+
+def create_schema():
+    logger.info("Begin: creating inaturalist schema in DB")
+    return sql_runner("00_create_schema.sql")
```

Comment:
Since all of these are calling the same function, we can actually just call these functions directly in the DAG! They're all called sequentially too, so one way we could set this up is as follows:

_In the provider script_
```python
SQL_RUN_ORDER = [
    "00_create_schema.sql",
    "01_photos.sql",
    "02_observations.sql",
    "03_taxa.sql",
    "04_observers.sql"
]

...

if __name__ == "__main__":
    for script in SQL_RUN_ORDER:
        # Pull out the action being taken from the filename
        action = script.split("_", 1)[1].split(".")[0]
        log.info(f"Begin: {action}")
        sql_runner(script)
    ingest_inaturalist_data()
```

_In the DAG definition file_

```python
from airflow.models.baseoperator import chain

with DAG(...) as dag:

    scripts = []

    # Create all the tasks and add them to a list
    for script in SQL_RUN_ORDER:
        # This could even be simplified if we wanted to remove the `XX_` prefix and just split on file extension instead
        action = script.split("_", 1)[1].split(".")[0]
        script_task = PythonOperator(
            task_id=action,
            python_callable=inaturalist.sql_runner,
            op_kwargs={
                "file_name": script,
            }
        )
        scripts.append(script_task)

    # The final ingestion task
    ingest_data = PythonOperator(
        task_id="ingest_data", python_callable=inaturalist.ingest_inaturalist_data
    )

    # Chain all the tasks as dependencies
    # https://airflow.apache.org/docs/apache-airflow/stable/concepts/dags.html#task-dependencies
    chain(*scripts)

    # Add the ingestion task to the end
    scripts[-1] >> ingest_data
```

There are other ways to simplify this, but doing something like this prevents us from having to define short and almost identical functions for each step!

Reply:
Thanks! I think this is taken care of [here](https://github.com/WordPress/openverse-catalog/blob/fe3d0c27142b9d45e304520fc37aa9088f335d99/openverse_catalog/dags/providers/provider_api_scripts/inaturalist.py#L166), though I do still have creating the schema as a separate task, rather than part of the task group below. Annoying? Better to make one task group?